### PR TITLE
Remove `Rvalue::CheckedBinaryOp`

### PR DESCRIPTION
**Pull request by scottmcm**
Zulip conversation: <https://rust-lang.zulipchat.com/#narrow/stream/189540-t-compiler.2Fwg-mir-opt/topic/intrinsics.20vs.20binop.2Funop/near/438729996>
cc @RalfJung 

While it's a draft,
r? ghost

### CODE REVIEW

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 1]
```diff
@@ -1449,14 +1443,23 @@ pub enum BinOp {
     Add,
     /// Like `Add`, but with UB on overflow.  (Integers only.)
     AddUnchecked,
+    /// Like `Add`, but returns `(T, bool)` of both the wrapped result
+    /// and a bool indicating that it overflowed.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:1447]:
```suggestion
    /// and a bool indicating whether it overflowed.
```
same for the other two

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 2]
```diff
@@ -1064,14 +1064,7 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
...
                     }
-                }
-            }
-            Rvalue::CheckedBinaryOp(op, vals) => {
-                use BinOp::*;
-                let a = vals.0.ty(&self.body.local_decls, self.tcx);
-                let b = vals.1.ty(&self.body.local_decls, self.tcx);
-                match op {
-                    Add | Sub | Mul => {
+                    AddWithOverflow | SubWithOverflow | MulWithOverflow => {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/validate.rs:1067]: Can this be merged with `AddUnchecked | SubUnchecked | MulUnchecked | Shl | ShlUnchecked | Shr
| ShrUnchecked`? They all have the same requirements I think -- having two ints.

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 3]
```diff
@@ -1088,7 +1081,6 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             );
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/validate.rs:1081]: The "unequal types" thing is already checked up above the `match`.

**Review by RalfJung — COMMENTED** [review]
Yes please :)

I only looked at the interpreter-related parts and the general MIR syntax.

I think further cleanup should be possible, e.g. in the interpreter I don't think we want to even still have `binop_{with,ignoring}_overflow`, and maybe codegen also can be simplified, but that can be done on follow-up PRs.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**